### DISCUSSION TITLE
Cannot export 'rpool': pool is busy

### DISCUSSION THREAD

**Discussion by spflanze** [openzfs/zfs#13779] [Q&A]
I am attempting to install Ubuntu on a ZFS encrypted, and RAIDed, file system. The latest version of the procedure I am attempting to do this with can be downloaded from:
https://www.mediafire.com/file/2w8mdb96tbzslub/Ubuntu_22.04_Root_on_ZFS_Encryption.odt/file
This procedure is a customized version of the one at:
https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html

At Step 5.10, I get an error in response to the command:
```
zpool export -a
cannot export 'rpool': pool is busy
```
I also tried the command "zpool export -f -a" and got the same result.

I tried making the present working directory one that is outside of the rpool, which was the /home directory, and this did not make a difference.
Also I tried closing the ssh connection . This too did not make a difference.

I tried the command sequence:
```
root@ubuntu:/home# zpool list
NAME    SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
rpool  1.73T  1.21G  1.73T        -         -     0%     0%  1.00x    ONLINE  /mnt
root@ubuntu:/home# fuser -k /mnt
root@ubuntu:/home# zpool export -f -a
cannot export 'rpool': pool is busy
```

How important is it that rpool be exported before proceeding? 
What is the best way to find out what is keeping rpool busy?
Any other suggestions about what to do about this error would be appreciated.

**Comment by spflanze** [comment]
I got Ubuntu installed. And updated procedure can be downloaded from the same link I opened this discussing with. I was not able to get rpool exported. Not being able to meant that I had to do some manual operations in BusyBox (aka intramfs) regarding rpool. BusyBox displayed the commands to do. Then I found it was not sufficient to use a reboot command. To get the minimal OS to boot, I had to cycle power. I used the "poweroff" command instead of "reboot".

**Comment by jstasiak** [comment]
#16036 describes the following way of figuring out which processes prevent the pool from being exported:

```
grep rpool /proc/*/mounts | cut -d/ -f3 | uniq | xargs ps
```

**Reply by colinstu12** [reply]
I get 'error: process ID list syntax error' when running this? Running it without the '| xargs ps' will then work but it returns a list of pids.

**Reply by jstasiak** [reply]
Interesting, can you show what's the exact output without the `| xargs ps` part?

**Reply by colinstu12** [reply]
```
root@pm:~# grep rpool /proc/*/mounts | cut -d/ -f3 | uniq
1004
100
1011
101
102
104
105
106
107
108
10
...
... a huge long list of numbers, cutting for brevity (no periods like I'm putting here)
...
87
88
8902
8903
89
9067
90
91
92
93
94
95
96
97
989
98
994
996
998
99
9
self
thread-self
root@pm:~# 
```

Is it not happy about the "self" and "thread-self"?

**Reply by jstasiak** [reply]
Yeah that'd be my suspicion. Not sure what these are, maybe the `cut` command needs to be adjusted or some data filtered out, depending on what exactly `grep rpool /proc/*/mounts` produces.

**Comment by bcsmith2846** [comment]
The command given by @jstasiak from #16036 gives the correct processes to kill. This seems to happen when using SSH to run the commands vs being physically at the server terminal. Once I killed all the processes (which killed the SSH session), I went to the physical server's terminal and ran `zpool export -a` and it succeeded.

**Comment by grahamperrin** [comment]
Ubuntu aside …

On FreeBSD:

- I use [lsof(8)](https://man.freebsd.org/cgi/man.cgi?query=lsof&sektion=8&manpath=freebsd-ports) to identify any process that prevents export
- in my case, killing is required only if gvfsd-trash is the culprit ([bug 254024](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=254024)).

[sysutils/lsof](https://www.freshports.org/sysutils/lsof/)
